chapters/cpp_for_opencl.md

```diff
@@ -12,8 +12,103 @@ C++ for OpenCL is supported by [Clang](https://clang.llvm.org/docs/UsersManual.h
+}
+int get_re() { return m_re; }
+int get_im() { return m_im; }
+};
+
+// A helper function to compute multiplication over
+// complex numbers read from the input buffer and
+// to store the computed result into the output buffer.
+template<typename T>
+void compute_helper(global T *in, global T *out) {
+  auto idx = get_global_id(0);	
+  // Every work-item uses 4 consecutive items from the input
+  // buffer - two for each complex number.
+  auto offset = idx * 4;
+  complex_t num1{in[offset], in[offset + 1]};
+  complex_t num2{in[offset + 2], in[offset + 3]};
+  // Perform complex number multiplication.
+  complex_t res = num1 * num2;
+  // Every work-item writes 2 consecutive items to the output
+  // buffer.
+  out[idx * 2] = res.get_re();
+  out[idx * 2 + 1] = res.get_im();
+}
+
+// This kernel can be used for complex number multiplication
+// in single precision.
+kernel void compute_sp(global float *in, global float *out) {
+  compute_helper(in, out);
+}
+
+// This kernel can be used for complex number multiplication
+// in half precision.
+#pragma OPENCL EXTENSION cl_khr_fp16: enable
+kernel void compute_hp(global half *in, global half *out) {
+  compute_helper(in, out); 
+}
+```
+
+## Developing kernels with C++ for OpenCL
+
+C++ for OpenCL sources can be developed and compiled just like OpenCL C sources. However due to the increased growth of application complexity especially those that benefit most from rich variety of C++ features and high-level abstractions, it is expected that the majority of C++ for OpenCL kernels will be compiled offline. The offline compilation also provides ability to take advantage of various newest features in open source tools and frameworks without waiting for their integration into the vendor toolchains. 
+
+### Offline compilation
+
+Clang provides support for C++ for OpenCL using regular interface as for any version of OpenCL C. More details can be found in its [UsersManual](https://clang.llvm.org/docs/UsersManual.html#cxx-for-opencl). In the majority of cases the generated binary can be used in the existing drivers. C++ for OpenCL version 1.0 is developed against OpenCL 2.0. Depending on the features used drivers from other versions (e.g. OpenCL 3.0) might be able to load the binaries produced with C++ for OpenCL v1.0 too. Use of global objects and static function objects with non-trivial constructors is not supported in a portable way, refer to [the following clang documentation](https://clang.llvm.org/docs/UsersManual.html#constructing-and-destroying-global-objects) for details.
+
+Clang only supports limited number of vendors and therefore to allow executing the binaries from C++ for OpenCL on more devices it is recommended to generate portable executable formants. C++ for OpenCL kernel sources can be compiled into SPIR-V using [open source tools](os_tooling.md) and then loaded into drivers supporting SPIR-V. C++ for OpenCL 1.0 generates SPIR-V 1.0 plus SPIR-V 1.2 where necessary.
```

Comment:
```suggestion
Clang only supports a limited number of vendors and therefore to allow executing the binaries from C++ for OpenCL on more devices it is recommended to generate portable executable formats. C++ for OpenCL kernel sources can be compiled into SPIR-V using [open source tools](os_tooling.md) and then loaded into drivers supporting SPIR-V. C++ for OpenCL 1.0 generates SPIR-V 1.0 plus SPIR-V 1.2 where necessary.
```

The last sentence isn't entirely correct:

> C++ for OpenCL 1.0 generates SPIR-V 1.0 plus SPIR-V 1.2 where necessary.

A language cannot generate SPIR-V, only tools can.

Reply:
Yeah indeed. I will update this thanks!